contracts/contract/deposit/RocketDepositPool.sol

```diff
@@ -150,28 +164,51 @@ contract RocketDepositPool is RocketBase, RocketDepositPoolInterface, RocketVaul
         uint256 balance = rocketVault.balanceOf("rocketDepositPool");
         uint256 totalEther = 0;
         // Calculate minipool assignments
-        uint256 maxAssignments = rocketDAOProtocolSettingsDeposit.getMaximumDepositAssignments();
-        MinipoolAssignment[] memory assignments = new MinipoolAssignment[](maxAssignments);
-        MinipoolDeposit depositType = MinipoolDeposit.None;
+
         uint256 count = 0;
         uint256 minipoolCapacity = 0;
-        for (uint256 i = 0; i < maxAssignments; ++i) {
-            // Optimised for multiple of the same deposit type
-            if (count == 0) {
-                (depositType, count) = rocketMinipoolQueue.getNextDeposit();
-                if (depositType == MinipoolDeposit.None) { break; }
-                minipoolCapacity = rocketDAOProtocolSettingsMinipool.getDepositUserAmount(depositType);
+        uint256 depositValueForAssignments = msg.value;
+        uint256 socializedAssignments = rocketDAOProtocolSettingsDeposit.getMaximumDepositAssignments();
+        MinipoolAssignment[] memory assignments = new MinipoolAssignment[](maxAssignments);
+
+        // Prepare half deposit assignments
+        count = rocketMinipoolQueue.getLength(MinipoolDeposit.Half);
+        minipoolCapacity = rocketDAOProtocolSettingsMinipool.getDepositUserAmount(MinipoolDeposit.Half);
+        for (uint256 i=0; i < count; ++i) { // (see note in full deposit loop)
+            if (depositValueForAssignments < minipoolCapacity) {
+                if (socializedAssignments == 0) { break; }
+                else {socializedAssignments--;}
+            } else {
+                depositValueForAssignments.sub(minipoolCapacity);
             }
-            count--;
-            if (minipoolCapacity == 0 || balance.sub(totalEther) < minipoolCapacity) { break; }
+            if (balance.sub(totalEther) < minipoolCapacity) { break; }
             // Dequeue the minipool
-            address minipoolAddress = rocketMinipoolQueue.dequeueMinipoolByDeposit(depositType);
+            address minipoolAddress = rocketMinipoolQueue.dequeueMinipoolByDeposit(MinipoolDeposit.Half);
             // Update running total
             totalEther = totalEther.add(minipoolCapacity);
             // Add assignment
             assignments[i].etherAssigned = minipoolCapacity;
             assignments[i].minipoolAddress = minipoolAddress;
         }
+
+        // Prepare full deposit assignments
+        count = rocketMinipoolQueue.getLength(MinipoolDeposit.Full);
+        minipoolCapacity = rocketDAOProtocolSettingsMinipool.getDepositUserAmount(MinipoolDeposit.Full);
+        for (i; i < i + count; ++i) { // NOTE - this is a weird line - we continue the indexing from the half deposit loop
```

Comment:
Wasn't sure if this weirdness was worth it. The alternative would be to have an assignment index variable and track/increment _that_ across both loops but let i reset. The current way saves a variable 🤷